### DISCUSSION TITLE
requests with cookies

### DISCUSSION THREAD

**Discussion by till** [mswjs/msw#1556] [Q&A]
Hey,

I have a weird situation where I started using MSW to mock out requests in tests (through the server implementation) and then tried to proceed to use the same mocks for my frontend (while developing the app).

There's a weird case where I do requests which are not mocked and MSW picks up (see it in the developer console) on them (letting me know about the request and suggests to add a handler). These requests are not supposed to be mocked and I want them to hit another backend.

Now on to the problem: it seems that the requests I don't want mocked hit my backend service, but they no longer include the cookie header anymore when they hit my backend. This is part of an SDK which uses Axios.

The problem is probably in my own code somewhere as I have been unable to re-produce it with an extracted test case, but I figured, I would ask here if there's anything obvious in the configuration of MSW (or otherwise), that would prevent cookies to work in these scenarios. Or if there's generally a problem with Axios as I saw a couple tickets.

For clarity:

 - http://ui.some.host - my app (mostly SPA, SvelteKit based)
 - http://api.some.host - my API service, I want these mocked
 - http://auth.some.host - auth service, I want these to proceed (always)

I guess, related to this, is there a way to tell MSW to ignore requests to a certain domain (or URL) so I don't see the messages at all? Or silence them?

After spending a few days on this, I started rebuilding the code in another branch and so far stayed away from the issue, but I am also not yet back to where I was mocking some of my frontend requests.

**Comment by kettanaito** [comment]
Hi, @till. Happy to hear from you.

> There's a weird case where I do requests which are not mocked and MSW picks up (see it in the developer console) on them (letting me know about the request and suggests to add a handler). These requests are not supposed to be mocked and I want them to hit another backend.

MSW warns on any requests that you haven't mocked to guard you against typos and unintended passthrough requests. You can disable that warning or have a custom logic around it using the [onUnhandledRequest](https://mswjs.io/docs/api/setup-worker/start#onunhandledrequest) option.

```js
worker.start({ onUnhandledRequest: 'bypass' })
```

You can also use this option to ignore bypassed requests to certain domains:

```js
worker.start({
  onUnhandledRequest(request, print) {
    // Never warn about unhandled requests sent to the "*.some.host" destination.
    if (request.url.hostname.endsWith('.some.host')) {
      return
    }

    print.warning()
  }
})
```

> Now on to the problem: it seems that the requests I don't want mocked hit my backend service, but they no longer include the cookie header anymore when they hit my backend.

I don't think this is MSW. We don't modify the bypassed requests in any way so your backend should receive exactly what you're sending from the client. 

You can triage this quite easily:

1. Disable MSW. Are the cookies sent now? If yes, please create a reproduction repo, and let's fix that together.
2. With MSW enabled, go to your "Network" tab in the browser DevTools, find the request with missing cookies, inspect it. Is the `Cookies` header there?
3. Double-check on your `credentials` and cookie security policies so you're actually allowed to send cookies across different hostnames.

**Reply by till** [reply]
Thanks for responding so fast. I'll have a look and will check it out as I rebuild. I didn't think it was MSW either. 😅

**Reply by rahul-keyValue** [reply]
@till Is the issue solved?

**Reply by till** [reply]
Can't tell. When I don't use the msw stuff, it works. Had no time to investigate further. But it's likely something local.

**Reply by rahul-keyValue** [reply]
@till I also got with the same issue, with 'bypass' the cookies are not attached to the requests.

**Reply by till** [reply]
@rahul-keyValue Maybe you can build an example case for someone to take a look?

**Reply by aristofun** [reply]
Something is breaking in MSW cookies passing for fastify in v 2.4.4 (2.4.3 works fine)

**Reply by kettanaito** [reply]
@aristofun, please upgrade and let me know. I've addressed a cookie-related issue in the latest releases. Hope that helps. If not, open a new issue with a reproduction repo. Thanks.

**Reply by aristofun** [reply]
Problem still there. I'm having hard time to make a reproducible repo, but 1 thing i noticed is that with 2.4.4+ version I see 'connection: close' in the response from requests made to localhost fastify server. That isn't the case with 2.4.3

**Reply by kettanaito** [reply]
MSW has no effect over those headers. If there's a change there, something else has caused it.

**Reply by aristofun** [reply]
yes, this smells like some dependency down the tree, but the gist of it is 
vitest test case with latext axios + https://www.npmjs.com/package/tough-cookie hitting latest fastify 5.* server started on localhost from the same vitest process - works fine with MSW 2.4.3, but not with 2.4.4+

I'll try to narrow it down when I have capacity.

**Reply by aristofun** [reply]
so far I remove Fastify from the equation - even raw node:http server doesn't receive cookie header in bypassed request

**Reply by aristofun** [reply]
Okay I finally created a minimal example: 
https://codesandbox.io/p/devbox/8qvlj6 

Change the msw version to `2.4.3` reinstall and `yarn start` to see the difference.

I don't think I have capacity nor expertise to take it further, but I would be truly grateful if you can fix it.

**Reply by aristofun** [reply]
https://github.com/mswjs/msw/issues/2338

**Reply by kettanaito** [reply]
Thanks for making a repo for it! Will take a look when I have a minute.
